### PR TITLE
05584 - AddressBookTestingTool app created

### LINKED ISSUE #5778: Create application to manually test AddressBookInitializer.
The following test cases for Address Book initialization should be manually testable with an new app. 
- [x] Whole network from genesis, swirld state should update address book.  
- [x] node restart, no upgrade, validate saved sign state used.
- [x] node restart on upgrade boundary, validate swirld state initialization of address book.
- [x] config flag for forcing use of config address book
- [ ] testing node with 0 stake updated behavior.

### PR DESCRIPTION
**Pull request by edward-swirldslabs**
**Description**:
This commit allows the manual testing of the AddressBookInitializer code through the AddressBookTestingTool.
* creates the AddressBookTestingTool project in platform-sdk/platform-apps/tests
* adds include of the AddressBookTestingTool project to settings.gradle.kts for platform-sdk
* updates config.txt with new app commented out.
* updates `Address.equalsWithoutStake(Address)` to use the byte encoding of the public keys and moves isOwnHost comparison to equals().
* adds logging for when the configuration address book is forced by settings to AddressBookInitializer.java

**Related issue(s)**:

Fixes #5778

**Notes for reviewer**:
The instructions for how to manually use the AddressBookTestingTool are in the file 
`platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md`

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (hendrikebbers) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:52]:
```suggestion
    private static final int TRANSACTIONS_PER_SECOND = 100;
```

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (cody-littley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:3]:
```suggestion
This document describes various manual tests that were performed using this testing app. This playbook should be used when designing automated tests using this app.
```

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (cody-littley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:48]: For automated tests we won't want to hard code this. But I'm assuming you are saving that for a follow up PR after you add the new workflow for supporting application configuration using the new configuration engine.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (cody-littley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:80]: The abstract class `PartialMerkleLeaf` has good default implementations for immutability. Use the `setImmutable()` method when you want to make this node immutable.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (cody-littley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:102]: use default method

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (cody-littley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:55]: Add `@NonNull` annotations to all non-primitive return types and parameters.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:62]: The bolded was not standing out in the text editor.   The whole sequence of tests should probably be done in order.  Some of the test are dependent on state setup by previous tests.   The term (changed) was indicating that the verification content had changed from the previous test.

Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:62]: If it is distracting, I can remove it.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:1]: I've changed it to README.md

I agree this is a stop-gap for not yet having automated testing capability.    That will hopefully be added soon. 

I created the volume of detail because the behavior is being handed off to another team and I wanted a walk-through of the complexity of behavior.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:55]: Prior job had the philosophy of coding explicitly to avoid ambiguity.  For example, always create the default constructor so that someone else doesn't wonder if it was left off by mistake.   Always have an else clause to an if statement.  Always comment inside empty blocks that the block is supposed to be empty, etc.

Is there a coding standard that I should be following?

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:62]: Added.  Had to add dependency to gradle build and module.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:90]: adding `@NonNull` and `Objects.requireNonNull()`

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java:38]: Added `@NonNull` and `Objects.requireNonNull()`

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:51]: fixed

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:22]: adding comments to contextualize the lines of code.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:62]: updated to (differs from previous section)

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:63]: replaced in all locations.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/AddressBookTestingToolInstructions.md:109]: updated

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java]: Once the Main can add its configuration in the browser, I will add the code to set the software version through querying the configuration value.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:80]: updated to setImmutable(true)

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:102]: removed overriding method.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:246]: Added markings in new commit.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:55]: added

File: platform-sdk/sdk/config.txt
Comment (edward-swirldslabs) [reply platform-sdk/sdk/config.txt:21]: Had to get creative to get two `\t` characters injected in each line, but it is done.

**Review by alittley — APPROVED** [review]

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/README.md
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/README.md:43]: I am way off my game.   My apologies for that.   fixed in the next commit.

Comment (alittley) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/README.md]: No worries 👌

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (cody-littley) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:64]: For serializable classes, a zero-arg constructor is a good idea even if it's empty, since it is required by the serialization framework. Adding another constructor in the future will cause runtime failures if the zero-arg constructor is missing.

For non-serializable classes, I think a zero-arg constructor that does nothing is not necessary.

**Review by cody-littley — DISMISSED** [review]

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java]: Adding logging statement, constructor is no longer empty.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/Address.java
Comment (alittley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/Address.java:822]: Is there a reason that `ownHost` is checked, here, rather than being added into `equalsWithoutStake`?

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (alittley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:57]: `platform` is marked `@NonNull`, but it is null until `init` is called

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (alittley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:84]: Is `stakingProfile` not being copied on purpose?

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java
Comment (alittley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java:43]: Primitives shouldn't be annotated

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java
Comment (alittley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java:46]: Primitive cannot be null, so no need to check

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (alittley) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:62]: Could be final (just noticed this now, feel free to ignore since it's late in the review process)

**Review by alittley — DISMISSED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/Address.java
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/address/Address.java:822]: I can't recreate the error in development that caused me to move it, so I moved it back.   It will cause problems if someone constructs a new address without checking the network to determine the isOwnHost value or if the node was in the set of local nodes to start.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolMain.java:57]: I removed the nullability annotations from the private member variables in the class.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:84]: I made it private final static.

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java:43]: fixed

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java
Comment (edward-swirldslabs) [reply platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/TransactionGenerator.java:46]: fixed

**Review by alittley — APPROVED** [review]

**Review by jeromy-cannon — APPROVED** [review]
from a release engineering perspective I approve the build changes, please wait for other approvers for remaining code before merging.

**Review by cody-littley — APPROVED** [review]

**Pull request merged**
